### PR TITLE
OSDOCS-7780: Porting RDS database with STS tutorial.

### PR DESCRIPTION
**Pull request by eohartman**
Version(s):
enterprise-4.13+

Issue:
[MOBB F2F Connect to RDS database with STS from ROSA](https://issues.redhat.com/browse/OSDOCS-7780)

Link to docs preview:
[Connect to RDS database with STS from ROSA](https://64937--docspreview.netlify.app/openshift-rosa/latest/cloud_experts_tutorials/cloud-experts-sts-rds-database-connect)

QE review:
- [ ] QE has approved this change.

Additional information:

### CODE REVIEW

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+== Environment Setup
+Configure the following environment variables.
+
+* Export cluster information:
++
+[source,terminal]
+----
+$ (`rosa list cluster`)
+   export CLUSTER_NAME=<cluster_name>
+   export AWS_REGION=$(rosa describe cluster -c ${ROSA_CLUSTER_NAME} -o json | jq -r .region.id)
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:46]:
```suggestion
   export AWS_REGION=$(rosa describe cluster -c ${CLUSTER_NAME} -o json | jq -r .region.id)
```

Comment (xueli181114) [reply rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:46]: Keep alignment with context

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+* The `jq` command
+
+== Environment Setup
+Configure the following environment variables.
+
+* Export cluster information:
++
+[source,terminal]
+----
+$ (`rosa list cluster`)
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:44]: This line is meaningless and will cause the command fail.
If we want to show customer how to find the <cluster name>, move it out from current command text to a single command

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+   export AWS_ACCOUNT_ID=$(aws sts get-caller-identity --query Account --output text)
+   export SCRATCH_DIR=/tmp/scratch
+   export AWS_PAGER=""
+   export PSQL_PASSWORD=$(openssl rand -base64 12)
+   export NODE=$(oc get nodes --selector=node-role.kubernetes.io/worker \
+    -o jsonpath='{.items[0].metadata.name}')
+   export VPC_ROSA=$(aws ec2 describe-instances \
+    --filters "Name=private-dns-name,Values=$NODE" \
+    --query 'Reservations[*].Instances[*].{VpcId:VpcId}' \
+    --region $AWS_REGION \
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:58]:
```suggestion
    --region $AWS_REGION --output json\
```

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    | jq -r '.[0][0].VpcId')
+   export ROSA_IP_OUT=$(aws ec2 describe-nat-gateways --filter "Name=vpc-id,Values=${VPC_ROSA}" --region ${AWS_REGION} \
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:60]:
```suggestion
   export ROSA_IP_OUT=$(aws ec2 describe-nat-gateways --filter "Name=vpc-id,Values=${VPC_ROSA}" --region ${AWS_REGION} --output json \
```

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+    | jq -r .NatGateways[].NatGatewayAddresses[].PublicIp)
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:61]:
```suggestion
    | jq -r '.NatGateways[].NatGatewayAddresses[].PublicIp')
```

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+   mkdir -p $SCRATCH_DIR
+----
+
+== Create database network
+
+. VPC and subnets:
++
+[source,terminal]
+----
+   $ VPC_DB=$(aws ec2 create-vpc --cidr-block 10.23.0.0/16 --region ${AWS_REGION} | jq -r .Vpc.VpcId)
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:71]:
```suggestion
   $ VPC_DB=$(aws ec2 create-vpc --cidr-block 10.23.0.0/16 --region ${AWS_REGION} --output json | jq -r .Vpc.VpcId)
```

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+   aws ec2 modify-vpc-attribute --vpc-id ${VPC_DB} --enable-dns-hostnames "{\"Value\":true}"
+   aws ec2 modify-vpc-attribute --vpc-id ${VPC_DB} --enable-dns-support "{\"Value\":true}"
+   SUBNET_A=$(aws ec2 create-subnet --vpc-id ${VPC_DB} --cidr-block 10.23.1.0/24 --availability-zone ${AWS_REGION}a | jq -r .Subnet.SubnetId)
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:74]:
```suggestion
   SUBNET_A=$(aws ec2 create-subnet --vpc-id ${VPC_DB} --cidr-block 10.23.1.0/24 --availability-zone ${AWS_REGION}a --output json | jq -r .Subnet.SubnetId)
```

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+   SUBNET_B=$(aws ec2 create-subnet --vpc-id ${VPC_DB} --cidr-block 10.23.2.0/24 --availability-zone ${AWS_REGION}b | jq -r .Subnet.SubnetId)
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:75]:
```suggestion
   SUBNET_B=$(aws ec2 create-subnet --vpc-id ${VPC_DB} --cidr-block 10.23.2.0/24 --availability-zone ${AWS_REGION}b --output json | jq -r .Subnet.SubnetId) 
```

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+. VPC and subnets:
++
+[source,terminal]
+----
+   $ VPC_DB=$(aws ec2 create-vpc --cidr-block 10.23.0.0/16 --region ${AWS_REGION} | jq -r .Vpc.VpcId)
+   aws ec2 modify-vpc-attribute --vpc-id ${VPC_DB} --enable-dns-hostnames "{\"Value\":true}"
+   aws ec2 modify-vpc-attribute --vpc-id ${VPC_DB} --enable-dns-support "{\"Value\":true}"
+   SUBNET_A=$(aws ec2 create-subnet --vpc-id ${VPC_DB} --cidr-block 10.23.1.0/24 --availability-zone ${AWS_REGION}a | jq -r .Subnet.SubnetId)
+   SUBNET_B=$(aws ec2 create-subnet --vpc-id ${VPC_DB} --cidr-block 10.23.2.0/24 --availability-zone ${AWS_REGION}b | jq -r .Subnet.SubnetId) 
+   SUBNET_C=$(aws ec2 create-subnet --vpc-id ${VPC_DB} --cidr-block 10.23.3.0/24 --availability-zone ${AWS_REGION}c | jq -r .Subnet.SubnetId)
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:76]:
```suggestion
   SUBNET_C=$(aws ec2 create-subnet --vpc-id ${VPC_DB} --cidr-block 10.23.3.0/24 --availability-zone ${AWS_REGION}c  --output json| jq -r .Subnet.SubnetId)
```

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+----
+
+. Internet gateway:
++
+[source,terminal]
+----
+  $ IGW=$(aws ec2 create-internet-gateway --region ${AWS_REGION} | jq -r .InternetGateway.InternetGatewayId)
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:83]:
```suggestion
  $ IGW=$(aws ec2 create-internet-gateway --region ${AWS_REGION}  --output json| jq -r .InternetGateway.InternetGatewayId)
```

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+----
+$ (`rosa list cluster`)
+   export CLUSTER_NAME=<cluster_name>
+   export AWS_REGION=$(rosa describe cluster -c ${ROSA_CLUSTER_NAME} -o json | jq -r .region.id)
+   export OIDC_PROVIDER=$(rosa describe cluster -c ${CLUSTER_NAME} -o json \
+    | jq -r .aws.sts.oidc_endpoint_url | sed -e 's/^https:\/\///')
+   export AWS_ACCOUNT_ID=$(aws sts get-caller-identity --query Account --output text)
+   export SCRATCH_DIR=/tmp/scratch
+   export AWS_PAGER=""
+   export PSQL_PASSWORD=$(openssl rand -base64 12)
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:52]: This line may not work. The random password may contain unusable characters. I got `IIbvew2/bTQsB5oT` first time which cannot work due to password validation `An error occurred (InvalidParameterValue) when calling the CreateDBInstance operation: The parameter MasterUserPassword is not a valid password. Only printable ASCII characters besides '/', '@', '"', ' ' may be used.`

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+       --db-instance-class db.t3.micro \
+       --engine postgres \
+       --master-user-password ${PSQL_PASSWORD} \
+       --allocated-storage 20 \
+       --master-username postgres \
+       --region ${AWS_REGION} \
+       --db-subnet-group-name db-group-${CLUSTER_NAME} \
+       --enable-iam-database-authentication \
+       --publicly-accessible \
+       --region ${AWS_REGION} \
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:116]: --output json

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+                   "rds-db:connect"
+               ],
+               "Resource": [
+                   "arn:aws:rds-db:${AWS_REGION}:${AWS_ACCOUNT_ID}:dbuser:$(echo ${RDS_DB} | jq -r .DbiResourceId)/iamuser"
+               ]
+           }
+       ]
+   }
+   EOF
+----
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:155]: The file generated by above command cannot be used for creation unless I format the content again with jsonformatter

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+. Create the RDS access policy:
+[NOTE] 
+This creates a named policy for the cluster. You can use a generic policy for multiple clusters.
++
+[source,terminal]
+----
+  $ POLICY=$(aws iam create-policy --policy-name "${CLUSTER_NAME}-rosa-rds-policy" \
+      --policy-document file://$SCRATCH_DIR/rds-policy.json \
+      --query 'Policy.Arn' --output text)
+   echo $POLICY
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:166]: Run the two command together will get empty value, unless I run them separatedly

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
++
+[source,terminal]
+----
+  $ curl -s -H "Accept: application/json" "https://sts.amazonaws.com/\
+   ?Action=AssumeRoleWithWebIdentity\
+   &DurationSeconds=3600\
+   &RoleSessionName=test\
+   &RoleArn=${ROLE}\
+   &WebIdentityToken=$(oc create token default --audience openshift --duration 60m)\
+   &Version=2011-06-15" | jq
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:234]: The copied command has whitespaces those caused command run failure

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+----
+  $ psql -h ${DB_ENDPOINT}
+   CREATE USER iamuser WITH LOGIN; 
+   GRANT rds_iam TO iamuser;
+   CREATE DATABASE iamdb;
+   \c iamdb
+   CREATE EXTENSION if not exists ip4r;
+   GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO iamuser;
+   quit
+----
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:259]: lack of `\q` to exit psql
lack of `exit` to quit pod

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 17]
```diff
@@ -0,0 +1,399 @@
...
+  $ psql -h ${DB_ENDPOINT} -U iamuser -d iamdb
+   CREATE TABLE if not exists ref_ip_blocks
+   (
+     iprange iprange,
+     geoname varchar
+   
+   );
+   \copy ref_ip_blocks FROM 'geolite2-country-ipv4.csv' DELIMITER ',' CSV;
+   CREATE INDEX ref_ip_blocks_ip4r_idx on ref_ip_blocks using gist(iprange);
+   quit
```
Comment (xueli181114) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:295]: lack of `\q` to quit psql

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 18]
```diff
@@ -0,0 +1,399 @@
...
+   exit
```
Comment (michaelryanmcneill) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:296]: Replace with:

```
$ psql -h ${DB_ENDPOINT} -U iamuser -d iamdb
$ CREATE TABLE if not exists ref_ip_blocks
$ (
  iprange iprange,
  geoname varchar

);
$ \copy ref_ip_blocks FROM 'geolite2-country-ipv4.csv' DELIMITER ',' CSV;
$ CREATE INDEX ref_ip_blocks_ip4r_idx on ref_ip_blocks using gist(iprange);
$ quit
$ exit
```

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 19]
```diff
@@ -0,0 +1,399 @@
...
+     name: aws-creds
+   type: Opaque 
+   stringData: 
+     credentials: |
+       [default]
+       role_arn = ${ROLE}
+       web_identity_token_file = /var/run/secrets/openshift/serviceaccount/token
+   EOF
+   
+   oc patch deployment ip-finder-api --type=merge -p '{"spec":{"template":{"spec":{"volumes":[{"name":"bound-sa-token","projected":{"sources":[{"serviceAccountToken":{"audience":"openshift","expirationSeconds":3600,"path":"token"}}]}},{"name":"aws-creds","secret":{"secretName":"aws-creds"}}]}}}}'
```
[the remark this answers is not in the document]
Comment (michaelryanmcneill) [reply rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:330]: The indentation is breaking this command. Replace with:

```
$ oc apply -f - <<EOF
apiVersion: v1
kind: Secret
metadata:
  name: aws-creds
type: Opaque 
stringData: 
  credentials: |
    [default]
    role_arn = ${ROLE}
    web_identity_token_file = /var/run/secrets/openshift/serviceaccount/token
EOF
$ oc patch deployment ip-finder-api --type=merge -p '{"spec":{"template":{"spec":{"volumes":[{"name":"bound-sa-token","projected":{"sources":[{"serviceAccountToken":{"audience":"openshift","expirationSeconds":3600,"path":"token"}}]}},{"name":"aws-creds","secret":{"secretName":"aws-creds"}}]}}}}'
$ oc patch deployment ip-finder-api --type='json' -p='[{"op": "add", "path": "/spec/template/spec/containers/0/volumeMounts", "value": [{"name":"bound-sa-token","readOnly":true,"mountPath":"/var/run/secrets/openshift/serviceaccount"},{"name":"aws-creds","mountPath":"/opt/app-root/src/.aws"}]}]'
```

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 20]
```diff
@@ -0,0 +1,399 @@
...
+   
+   oc patch deployment ip-finder-api --type='json' -p='[{"op": "add", "path": "/spec/template/spec/containers/0/volumeMounts", "value": [{"name":"bound-sa-token","readOnly":true,"mountPath":"/var/run/secrets/openshift/serviceaccount"},{"name":"aws-creds","mountPath":"/opt/app-root/src/.aws"}]}]'
+----
+
+. Expose app:
++
+[source,terminal]
+----
+  $ oc expose service ip-finder-api
+   oc patch route ip-finder-api --patch '{"spec":{"tls":{"termination":"edge","insecureEdgeTerminationPolicy":"Redirect"}}}'
```
Comment (michaelryanmcneill) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:340]: Broken; replace with:

```
$ oc expose service ip-finder-api
$ oc patch route ip-finder-api --patch '{"spec":{"tls":{"termination":"edge","insecureEdgeTerminationPolicy":"Redirect"}}}'
```

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 21]
```diff
@@ -0,0 +1,399 @@
...
+  $ curl https://$(oc get route ip-finder-api -o jsonpath='{.spec.host}')
+----
+Expected output:
++
+[source,terminal]
+----
+   {
+    "your_ip": X.X.X.X",
+    "countrycode": "FR"
+   }
```
[the remark this answers is not in the document]
Comment (michaelryanmcneill) [reply rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:356]: With my changes this error is not detected.

File: rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc [hunk 22]
```diff
@@ -0,0 +1,399 @@
...
+----
+  $ oc delete ns rds-sts-app
+   aws rds delete-db-instance --db-instance-identifier psql-${CLUSTER_NAME} --region ${AWS_REGION} --skip-final-snapshot
+   aws ec2 detach-internet-gateway --vpc-id ${VPC_DB} --internet-gateway-id ${IGW}
+   aws ec2 delete-internet-gateway --internet-gateway-id ${IGW} --region ${AWS_REGION}
+   aws ec2 delete-subnet --subnet-id ${SUBNET_A} --region ${AWS_REGION}
+   aws ec2 delete-subnet --subnet-id ${SUBNET_B} --region ${AWS_REGION}
+   aws ec2 delete-subnet --subnet-id ${SUBNET_C} --region ${AWS_REGION}
+   aws ec2 delete-vpc --vpc-id ${VPC_DB} --region ${AWS_REGION}
+   aws rds delete-db-subnet-group --db-subnet-group-name db-group-${CLUSTER_NAME}
```
Comment (michaelryanmcneill) [inline rosa_tutorials/rosa-mobb-sts-rds-database-connect.adoc:373]: Replace:

```
$ oc delete ns rds-sts-app
$ aws rds delete-db-instance --db-instance-identifier psql-${CLUSTER_NAME} --region ${AWS_REGION} --skip-final-snapshot
$ aws ec2 detach-internet-gateway --vpc-id ${VPC_DB} --internet-gateway-id ${IGW}
$ aws ec2 delete-internet-gateway --internet-gateway-id ${IGW} --region ${AWS_REGION}
$ aws ec2 delete-subnet --subnet-id ${SUBNET_A} --region ${AWS_REGION}
$ aws ec2 delete-subnet --subnet-id ${SUBNET_B} --region ${AWS_REGION}
$ aws ec2 delete-subnet --subnet-id ${SUBNET_C} --region ${AWS_REGION}
$ aws ec2 delete-vpc --vpc-id ${VPC_DB} --region ${AWS_REGION}
$ aws rds delete-db-subnet-group --db-subnet-group-name db-group-${CLUSTER_NAME}
```

**Review by michaelryanmcneill — CHANGES_REQUESTED** [review]
Indentation is causing severe issues across all of the code blocks. In addition to the identified changes, please ensure that you fix any indentation in code blocks that is unintentional.

**Comment by eohartman** [comment]
@michaelryanmcneill @xueli181114 Can y'all please take a look at this and let me know what you think? 

Note: There are still some formatting issues under the "IAM Permissions" section, regarding step numbering and indentation of a code block. I'm still working on that, but wanted to get y'all to review the content to ensure the rest of the requested changes have been resolved.

Thank you!

**Pull request closed without merging**